proposals/nnnn-rpc-design-refactoring.md

```diff
@@ -0,0 +1,153 @@
+* Status: **Awaiting review**
+* Impacted Platforms: [Core]
+
+## Introduction
+
+This proposal is to rework design of application manager layer in order to:
+ - Reduce coupling between components
+ - Reduce responsibility of ApplicationManager class
+ - Provide plugin architecture for easy adding additional functionality and RPC groups   
+
+## Motivation
+
+SDL core project is actively implementing new features and new functionality.  
+To keep project on track and to reduce risks of regression during implementation
+refactoring of application manager layer should be done.  
+
+Currently, extending SDL functionality may require code duplication. 
+RC module, in fact, duplicates logic of application_manager layer, there is separate request controller, separate mechanism to call  polcies checks, and RPC's processing. So any common logic that should be implemented in application_manager layer should be duplicated in request controller, it results in bugs, and requires more time for implementation. 
+
+The following steps will provide ability to keep RC enough encapsulated, without duplicating SDL code :
+
+## Proposed solution
+#### Reduce coupling between components and use dependency injection practice
+
+Now Application manager is a service locator and should to be passed to each component
+even if component does not need application manager, but need some other classes that owned by application manager.
+During construction components should explicitly receive all dependencies. 
+It will prevent from unwanted access and improve encapsulation. 
+
+
+#### Reduce responsibility of ApplicationManager class 
+Currently application manager class has too many responsibilities:
+ - Service location
+ - Mobile/HMI RPC processing 
+ - Mobile/HMI RPC sending 
+ - Registration and managing mobile applications
+ - Handle different states of SDL (low voltage/ video streaming / audio_path_throw ...) 
+ 
+ The purpose of this refactoring is to extract handling/sending RPC's to separate module.  
+ Many SDL commands require application manager only for sending RPC's and do not need to know anything related to other application manager responsibility. 
+ Also, extracting RPC processing to separate module will reduce regression risk of changing any functionality that now in ApplicationManager.
+ 
+#### Provide plugin architecture for easy adding additional functionality and RPC groups   
+
+SDL RPC's can be splitted to some groups or RPC's related to certain functionality (for example RC or video streaming).  
+Usually these RPC's work with some functionality that is not required for other RPC's, so should not be exposed. 
+
+Adding new RPC or group of RPC's is now very complicated process, but it is also not very obvious. 
+If OEM vendors need to introduce in their custom SDL forks new functionality, they need to change SDL code, 
+add new classes and modify existing factories. 
+
+The best solution would be plugin approach of adding new RPC's.
```

Comment:
The best solution is to adopt a plugin approach of adding new RPC's to SDL core.

Reply:
@mrapitis, done in 2c9f7a3f7bb7e6c8b26bb1e4aec779cf65733666